backend/result/score.py

```diff
@@ -1,74 +1,105 @@
     choices = list(data['choices'].keys())
     return choices.index(data['value']) + 1
 
 
-def reaction_time_score(result, data):
-    expected_response = check_expected_response(result)
-    json_data = result.json_data
+def reaction_time_score(result: Result, data: ScoringData) -> float:
+    """Return the difference between the configured maximal `response_time`
+    and the participant's reaction time (`decision_time`)
+    If the answer of the participant is incorrect, return the negative reaction time
+    """
+    expected_response = result.expected_response
+    json_data = (
+        result.json_data
+    )  # TODO: raise exception if either expected_response or json_data is `None`
     if expected_response and json_data:
-        time = json_data.get('decision_time')
-        timeout = json_data.get('config').get('response_time')
+        time = json_data.get(
+            'decision_time'
+        )  # TODO: raise exception if json_data does not contain decision_time
+        timeout = json_data.get('config').get(
+            'response_time'
+        )  # TODO: raise exception if json_data does not contain config with response_time
         if expected_response == data['value']:
             return math.ceil(timeout - time)
         else:
             return math.floor(-time)
 
 
-def song_sync_recognition_score(result, data):
+def song_sync_recognition_score(result: Result, data: ScoringData) -> float:
+    """First step of SongSync scoring (used in experiment.rules.hooked and derivatives):
+    if the participant gives no (timeout) or a negative response to 'Do you know this song?', return 0
+    otherwise, return the decision time
+    """
     if result.given_response == 'TIMEOUT' or result.given_response == 'no':
         return 0
     json_data = result.json_data
     if json_data:
-        time = json_data.get('decision_time')
+        time = json_data.get(
+            'decision_time'
+        )  # TODO: raise exception if time or timeout are `None`
         timeout = json_data.get('config').get('response_time')
         return math.ceil(timeout - time)
 
 
-def song_sync_continuation_score(result, data):
-    ''' modify previous result and return None'''
+def song_sync_continuation_score(result: Result, data: dict):
+    """Second step of the SongSync scoring, only happens if recognition score was nonzero.
+    After continuation of audio, participants need to decide whether it was continued in the correct place.
+    If answered incorrectly, this function modifies the reaction time score of the previous step.
+    """
     previous_result = result.session.last_result(["recognize"])
-    if check_expected_response(result) != result.given_response:
-        previous_result.score *= -1
+    if result.expected_response != result.given_response:
+        previous_result.score *= (
+            -1
+        )  # will raise exception if `previous_result.score = None`
```

Comment:
Should this be a TODO: as well then?

Reply:
Hm, perhaps it would be good to catch the exception and log a message about the special conditions for this scoring rule.